feat: improve tuist graph command

alexanderwe opened the pull request:
### Short description 📝

This PR should resolve two things about the `graph` command:

1. Do not output unnecessary log messages 
2. If no output path is supplied on `--format json` the stdout should be used to print the JSON 

### How to test the changes locally 🧐

tbd

### Contributor checklist ✅

- [ ] The code has been linted using run `mise run lint-fix`
- [ ] The change is tested via unit testing or acceptance testing, or both
- [ ] The title of the PR is formulated in a way that is usable as a changelog entry
- [ ] In case the PR introduces changes that affect users, the documentation has been updated

### Reviewer checklist ✅

- [ ] The code architecture and patterns are consistent with the rest of the codebase
- [ ] Reviewer has checked that, if needed, the PR includes the label `changelog:added`, `changelog:fixed`, or `changelog:changed`, and the title is usable as a changelog entry

alexanderwe commented:
@pepicrft @fortmarek I have digged into the code and tried to get an overview about the current logging situation (not yet touched the output path). I saw that in total there are currently 3 ways how something can be printed on stdout: 

1. The logs controller 
2. The UI context 
3. The alert context 

I experimented a little bit with the quiet mode. I am not sure if this is the right approach here. I would therefore like to get a better understanding what's your vision for the "output heaviness" of tuist. I can see that nice UI output is primarily focused for human interaction. 

Would agree that by using the `--quiet` flag, we should completely prevent any output or should `--quiet` only prevent informational and "ui" output but still showing errors ?  
If we are doing the latter, I think this information than has to be injected into the `loggerHandler`, `ui` and `alertsController` as well. 

Please correct me if I am on the complete wrong path here. Looking forward to your thoughts.

alexanderwe commented:
> > Would agree that by using the --quiet flag, we should completely prevent any output or should --quiet only prevent informational and "ui" output but still showing errors ?
> 
> `--quiet` should prevent any output with the exception of errors, yeah.
> 
> > If we are doing the latter, I think this information than has to be injected into the loggerHandler, ui and alertsController as well.
> 
> Yes, you are correct. I'd expect `loggerHandler` to have that information already injected, bu the new `ui` and `alertsController` might not cc @pepicrft

Thanks a lot for the input! I focused on making the `GraphCommand` output to stdout now. I will continue to look into how to incorporate the `quiet` information to the `ui` and `alertsController`. 

While `alertsController` is already an abstraction, the `ui` is directly equivalent to the `Noora` type. In order to inject the `quiet` I probably need to wrap `Noora` into another type to prevent the output. Is that fine from your side ? 
In addition, what is the standard way to print something to stdout in tuist ? I couldn't find a direct method other than by using the logger instance. Is this the recommended way to?

fortmarek commented:
> I probably need to wrap Noora into another type to prevent the output. Is that fine from your side ?

I think that's reasonable, yeah.

> I couldn't find a direct method other than by using the logger instance. Is this the recommended way to?

Yes, I think for the `json` output, we should use `logger` for now. At some point, we should move to Noora, but we currently don't have a method for outputing a raw text. If you're gonna build a `ui` abstraction, I'd make the sure the JSON output is done via the `ui` abstraction, so it's easier for us to change this down the line.

alexanderwe commented:
@fortmarek I have introduced a `UIController` which is a simple proxy for Noora at the moment. I also introduced a `message` method which allows to simply print a message formatted with Noora. 

In addition to just allow `json` to be printed to stdout I also did the same for `svg` since it is a simple text as well and could potentially also just be printed out. 

A few issues I found while doing that: 

1. Noora does not offer a formatted method like mentioned here: https://noora.tuist.dev/text-styling#usage so I hardcoded theme and terminal to the default values 
2. There is no way in Noora to suppress the error messages of the `progressStep` so these I could not influence by ``--quiet``
3. By using the `logger` I was not able to enable logging with ``--quiet`` attached, so I opted to ``print``

I think there might be some further changes necessary to the overall CLI usage if we go with allowing _some_ output if ``--quiet`` is supplied.

pepicrft commented on Sources/TuistKit/Commands/GraphCommand.swift:
```diff
@@ -81,7 +79,15 @@ public struct GraphCommand: AsyncParsableCommand {
+        // If output path is present
+        var absoluteOutputPath: AbsolutePath?
+        if let outputPath {
+            absoluteOutputPath = try AbsolutePath(validating: outputPath, relativeTo: FileHandler.shared.currentPath)
```
I'd use [this one](https://github.com/tuist/FileSystem/blob/1bff6d54e90a79706a4a9d77cd081104de430361/Sources/FileSystem/FileSystem.swift#L255) creating an instance of scoped to the command (i.e. a property).

alexanderwe replied:
You mean like this: https://github.com/tuist/tuist/pull/7363/commits/9a8698aad019755b4489656bcf77660ab8435eaa#diff-b3d112961f66eeb993eb40267f3b8e91015a3e1a9747f240732e4a4291e0c3f2R91 ?

pepicrft commented on Sources/TuistKit/Commands/GraphCommand.swift:
```diff
@@ -102,6 +106,18 @@ enum GraphFormat: String, ExpressibleByArgument, CaseIterable {
+    /// Flag to indicate if an output to stdout is allowed.
+    var allowsStdOut: Bool {
+        switch self {
+        case .json, .svg, .dot:
+            return true
+        default:
+            return false
```
Agreed

alexanderwe replied:
Covered all enum cases explicitly now.

pepicrft commented on Sources/TuistKit/Commands/GraphCommand.swift:
```diff
@@ -102,6 +106,18 @@ enum GraphFormat: String, ExpressibleByArgument, CaseIterable {
     case dot, json, legacyJSON, png, svg
 }
 
+extension GraphFormat {
```
Nit: Since this is only used by the service, I'd move it there and make it private.

alexanderwe replied:
Moved it and made it private

pepicrft commented on Sources/TuistKit/Services/GraphService.swift:
```diff
@@ -60,15 +60,118 @@ final class GraphService {
-        outputPath: AbsolutePath
+        outputPath: AbsolutePath?
     ) async throws {
+        // Sanity checks
```
```suggestion
```

alexanderwe replied:
Removed

pepicrft commented on Sources/TuistKit/Commands/GraphCommand.swift:
```diff
@@ -102,6 +106,18 @@ enum GraphFormat: String, ExpressibleByArgument, CaseIterable {
 
+extension GraphFormat {
+    /// Flag to indicate if an output to stdout is allowed.
+    var allowsStdOut: Bool {
```
I'd name this one:
```suggestion
    var isTextPrintable: Bool {
```

alexanderwe replied:
Changed

pepicrft commented on Sources/TuistKit/Services/ProjectAutomation+ManifestMapper.swift:
```diff
@@ -289,13 +289,19 @@ extension ProjectAutomation.BuildConfiguration.Variant {
 enum GraphServiceError: FatalError {
```
Since you are touching this one, would you mind updating it to `LocalizedError` and turn `description` into `errorDescription: String?`? We are moving away from using `FatalError`

pepicrft replied:
You can then drop `type`.

alexanderwe replied:
Sure, no problem. Like the change 👍

pepicrft commented on Sources/TuistSupport/UI/UIController.swift:
```diff
@@ -0,0 +1,117 @@
+import Noora
+
+/// A controller for UI elements on the terminal.
+public struct UIController: Noorable {
```
This one is extending by wrapping to:
- Add a new capability (stateless raw output).
- Conditionally output based on an attribute.

I think those two things are candidates to be capabilities of [Noora](https://github.com/tuist/noora):

```swift
let noora = Noora(quiet: true)
noora.raw("\(jsonGraph)")
```

> [!NOTE]
> Nora does internal logging even in the case of `quiet: true`, which is useful for debugging purposes.

alexanderwe replied:
I agree, it would be nice to have this functionality directly available in Noora. Then the controller becomes obsolete. I added some TODOs in the file,  to not loose sight of the upcoming changes, is this ok ?

fortmarek replied:
@alexanderwe should we instead implement first the changes in Noora before going ahead with this PR? Or do you think the `UIController` would be needed even with those changes in Noora?

alexanderwe replied:
@fortmarek As soon as Noora provides the possibility to be "quiet" and has a "raw" output format I think it is not necessary to have the `UIController` wrapper. 

If we add the `isQuiet` property to Noora I think we also need a way to override it. e.g. for the case where we want to raw output the JSON but still everything else should be quiet.

alexanderwe replied:
In terms of if we should first add the changes to Noora depends a little bit on when you would like to see this changes land in `main` 😄

fortmarek replied:
We're not in a rush. I'd lean to first doing the changes in Noora and then going back to this PR. What do you think, @alexanderwe?

alexanderwe replied:
Then let's wait until the changes land in Noora, then the complexity of the codebase is not increased and the number of LOC can be reduced here.

pepicrft commented on Sources/TuistSupport/UI/UIController.swift:
```diff
@@ -0,0 +1,117 @@
+    let noora: Noorable
+
+    /// Flag to silent all output
+    let isQuiet: Bool = ProcessInfo.processInfo.environment[Constants.EnvironmentVariables.quiet] != nil
```
Once we move the responsibility to Noora, this information should be passed in `TuistCommand` when instance of `Noora` is created.

alexanderwe replied:
Added a TODO for this as well

pepicrft reviewed (commented):
Thanks a lot for working on this one @alexanderwe 🙏🏼. You did an amazing job.
I left some comments regarding a different approach that we can take with the implementation.

alexanderwe commented:
Hi @fortmarek, @pepicrft thanks a lot for reviewing! I will check out your comments and implement the suggested changes asap. I will probably have time by the weekend, would this still fit your timelines?

alexanderwe commented:
> Hey @alexanderwe, does this also solve for the case when `--json` is passed, no other output apart from the JSON itself is sent to the stdout?
> 
> Otherwise, I'm aligned, thanks a lot for the contribution!

@fortmarek From my point of view yes. If you run  `tuist graph --format json --quiet` without specifying an output path the generated graph gets printed on stdout. Is this what you meant ?

alexanderwe commented on mise.toml:
```diff
@@ -4,7 +4,7 @@ swiftlint = "0.54.0"
 pnpm = "8.15.6"
 node = "20.12.0"
-sourcedocs = "2.0.2"
+#sourcedocs = "2.0.2"
```
I reverted the change. Locally I have issues with this tool when I run `mise install`.

```
 ~/.local/share/mise/plugins/sourcedocs/bin/download failed
* Downloading sourcedocs release 2.0.2...
curl: (56) The requested URL returned error: 404
asdf-sourcedocs: Could not download https://github.com/SourceDocs/SourceDocs/archive/refs/tags/2.0.2.zip
```

And then I cannot run the `lint-fix` command because `mise` first wants to download all tools. So I commented it in order to run the linter and formatter task.

alexanderwe commented on Sources/TuistSupport/UI/UIController.swift:
```diff
@@ -0,0 +1,117 @@
+                showSpinner: showSpinner,
+                task: task
+            )
+        }
```
I agree, I like it more. Not sure why I went with `if` in the first place. Changed all methods.

alexanderwe commented on Sources/TuistKit/Services/GraphService.swift:
```diff
@@ -60,15 +60,118 @@ final class GraphService {
+            guard let string = String(data: imageData, encoding: .utf8) else {
+                throw GraphServiceError.encodingError(GraphFormat.svg.rawValue)
+            }
+            ServiceContext.current?.logger?.notice("\(string)")
```
Changed, it uses the `ui` now

alexanderwe commented on Sources/TuistKit/Services/GraphService.swift:
```diff
@@ -60,15 +60,118 @@ final class GraphService {
+
+    // MARK: GraphViz Export
+
+    private func exportGraphVizToStdOut(
```
Changed the methods to `printGraphViz` and `printJSON`
